File: archivist/src/test/java/com/zorroa/archivist/repository/FolderDaoTests.java
package com.zorroa.archivist.repository;

import com.google.common.collect.ImmutableList;
import com.google.common.collect.Lists;
import com.zorroa.archivist.AbstractTest;
import com.zorroa.archivist.domain.*;
import com.zorroa.security.Groups;
import com.zorroa.archivist.security.UtilsKt;
import com.zorroa.archivist.service.PermissionService;
import com.zorroa.sdk.domain.Access;
import org.junit.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.dao.DataIntegrityViolationException;

import java.io.IOException;
import java.util.List;
import java.util.UUID;

import static com.zorroa.archivist.domain.FolderKt.getRootFolderId;
import static org.junit.Assert.*;

public class FolderDaoTests extends AbstractTest {

    @Autowired
    TaxonomyDao taxonomyDao;

    @Autowired
    FolderDao folderDao;

    @Autowired
    DyHierarchyDao dyHierarchyDao;

    @Autowired
    PermissionService permissionService;

    @Test
    public void testCreateAndGet()  {
        String name = "Foobar the folder";
        FolderSpec builder = new FolderSpec(name);
        Folder folder1 = folderDao.create(builder);

        Folder folder2 = folderDao.get(folder1.getId());
        assertEquals(folder2.getName(), name);
    }

    @Test
    public void testChildCount() throws IOException {
        String name = "Foobar the folder";
        FolderSpec builder1 = new FolderSpec(name);
        Folder folder1 = folderDao.create(builder1);
        assertEquals(0, folder1.getChildCount());

        FolderSpec builder2 = new FolderSpec(name + "2", folder1);
        Folder folder2 = folderDao.create(builder2);
        folder1 = folderDao.get(folder1.getId());
        assertEquals(1, folder1.getChildCount());

        folderDao.delete(folder2);
        folder1 = folderDao.get(folder1.getId());
        assertEquals(0, folder1.getChildCount());
    }

    @Test
    public void testGetByParentAndName() throws IOException {
        String name = "test";
        FolderSpec builder = new FolderSpec(name);
        Folder folder1 = folderDao.create(builder);
        Folder folder2 = folderDao.get(getRootFolderId(), name, false);
        assertEquals(folder1, folder1);
    }

    @Test
    public void testGetRootFolder() throws IOException {
        Folder folder2 = folderDao.get(getRootFolderId());
        assertNull(folder2.getParentId());
    }

    @Test
    public void testGetAll() throws IOException {
        Folder folder1 = folderDao.create(new FolderSpec("test1"));
        Folder folder2 = folderDao.create(new FolderSpec("test2"));
        Folder folder3 = folderDao.create(new FolderSpec("test3"));
        List<Folder> some = folderDao.getAll(
                Lists.newArrayList(folder2.getId(), folder3.getId()));
        assertTrue(some.contains(folder2));
        assertTrue(some.contains(folder3));
        assertFalse(some.contains(folder1));
    }

    @Test(expected=DataIntegrityViolationException.class)
    public void testCreateDuplicate() throws IOException {
        String name = "Foobar the folder";
        FolderSpec builder = new FolderSpec(name);
        folderDao.create(builder);
        folderDao.create(builder);
    }

    @Test
    public void testGetChildren() {
        String name = "Grandpa";
        FolderSpec builder = new FolderSpec(name);
        Folder grandpa = folderDao.create(builder);
        builder = new FolderSpec("Dad", grandpa);
        Folder dad = folderDao.create(builder);
        builder = new FolderSpec("Uncle", grandpa);
        Folder uncle = folderDao.create(builder);
        builder = new FolderSpec("Child", dad);
        Folder child = folderDao.create(builder);

        List<Folder> folders = folderDao.getChildren(grandpa);
        assertEquals(2, folders.size());
    }

    @Test
    public void testsetDyHierarchyRoot() {
        String name = "Foobar the folder";
        FolderSpec builder = new FolderSpec(name);
        Folder folder = folderDao.create(builder);
        assertTrue(folderDao.setDyHierarchyRoot(folder, "source.extension"));
    }

    @Test
    public void testSetTaxonomyRoot() {
        String name = "Foobar the folder";
        FolderSpec builder = new FolderSpec(name);
        Folder folder = folderDao.create(builder);

        Taxonomy tax = taxonomyDao.create(new TaxonomySpec(folder));
        assertTrue(folderDao.setTaxonomyRoot(folder, true));
        assertFalse(folderDao.setTaxonomyRoot(folder, true));

        assertTrue(folderDao.setTaxonomyRoot(folder, false));
        assertFalse(folderDao.setTaxonomyRoot(folder, false));
    }

    @Test
    public void removeDyHierarchyRoot() {
        String name = "Foobar the folder";
        FolderSpec builder = new FolderSpec(name);
        Folder folder = folderDao.create(builder);
        assertTrue(folderDao.setDyHierarchyRoot(folder, "source.extension"));
        assertTrue(folderDao.removeDyHierarchyRoot(folder));
    }


    @Test
    public void testGetChildrenInsecure() {
        authenticate("user");
        assertFalse(UtilsKt.hasPermission(Groups.ADMIN));

        Folder pub = folderDao.get(getRootFolderId(), "Users", false);
        int startCount =  folderDao.getChildren(pub.getId()).size();
        Folder f1 = folderDao.create(new FolderSpec("level1", pub));
        Folder f2 = folderDao.create(new FolderSpec("level2", pub));
        Folder f3 = folderDao.create(new FolderSpec("level3", pub));
        folderDao.setAcl(f3.getId(), new Acl().addEntry(
                permissionService.getPermission(Groups.ADMIN)));

        assertFalse(folderDao.hasAccess(f3, Access.Read));
        assertEquals(6, folderDao.getChildrenInsecure(pub.getId()).size());
        assertEquals(startCount+2, folderDao.getChildren(pub.getId()).size());
    }

    @Test
    public void testHasAccess() throws IOException {
        authenticate("user");
        FolderSpec builder = new FolderSpec("test");
        Folder folder1 = folderDao.create(builder);
        assertTrue(folderDao.hasAccess(folder1, Access.Read));

        Permission p = permissionService.createPermission(new PermissionSpec("group", "foo"));
        Acl acl = new Acl();
        acl.addEntry(p, Access.Read);
        folderDao.setAcl(folder1.getId(), acl);

        assertFalse(folderDao.hasAccess(folder1, Access.Read));
    }

    @Test
    public void testGetAndSetAcl() throws IOException {
        FolderSpec builder = new FolderSpec("test");
        Folder folder1 = folderDao.create(builder);

        Permission p = permissionService.createPermission(new PermissionSpec("group","foo"));
        folderDao.setAcl(folder1.getId(), new Acl().addEntry(p, Access.Read));

        Acl acl = folderDao.getAcl(folder1.getId());
        assertTrue(acl.hasAccess(p.getId(), Access.Read));
    }

    @Test
    public void testGetAndUpdateAcl() throws IOException {
        FolderSpec builder = new FolderSpec("test");
        Folder folder1 = folderDao.create(builder);

        Permission p1 = permissionService.createPermission(new PermissionSpec("group","foo"));
        folderDao.setAcl(folder1.getId(), new Acl().addEntry(p1, Access.Read));

        Acl acl = folderDao.getAcl(folder1.getId());
        assertTrue(acl.hasAccess(p1.getId(), Access.Read));

        Permission p2 = permissionService.createPermission(new PermissionSpec("group","bar"));
        folderDao.updateAcl(folder1.getId(), new Acl().addEntry(p2, Access.Read));

        acl = folderDao.getAcl(folder1.getId());
        assertTrue(acl.hasAccess(p1.getId(), Access.Read));
        assertTrue(acl.hasAccess(p2.getId(), Access.Read));
    }

    @Test
    public void testUpdateExitingAcl() throws IOException {
        FolderSpec builder = new FolderSpec("test");
        Folder folder1 = folderDao.create(builder);

        Permission p1 = permissionService.createPermission(new PermissionSpec("group","foo"));
        folderDao.setAcl(folder1.getId(), new Acl().addEntry(p1, Access.Read));
        folderDao.updateAcl(folder1.getId(), new Acl().addEntry(p1, Access.Read, Access.Write));

        Acl acl = folderDao.getAcl(folder1.getId());
        assertTrue(acl.hasAccess(p1.getId(), Access.Read));
        assertTrue(acl.hasAccess(p1.getId(), Access.Write));
        assertFalse(acl.hasAccess(p1.getId(), Access.Export));
    }

    @Test(expected=IllegalArgumentException.class)
    public void testGetAndSetBadAcl() throws IOException {
        FolderSpec builder = new FolderSpec("test");
        Folder folder1 = folderDao.create(builder);

        Permission p = permissionService.createPermission(new PermissionSpec("group","foo"));
        Acl acl = new Acl();
        acl.add(new AclEntry(p.getId(), 12));
        folderDao.setAcl(folder1.getId(), acl);
    }

    @Test
    public void testUpdate() {
        FolderSpec spec = new FolderSpec("v1");
        Folder v1 = folderDao.create(spec);

        FolderUpdate up = new FolderUpdate(v1);
        up.setName("v2");

        folderDao.update(v1.getId(), up);
        Folder v2 = folderDao.get(v1.getId());
        assertEquals("v2", v2.getName());
        assertEquals(null, v2.getSearch());
        assertEquals(0, v2.getAcl().size());
    }

    @Test
    public void testUpdateWithMove() {
        FolderSpec spec1 = new FolderSpec("f1");
        Folder f1 = folderDao.create(spec1);

        FolderSpec spec2 = new FolderSpec("f2");
        Folder f2 = folderDao.create(spec2);

        Folder root = folderDao.get(getRootFolderId());
        int rootCount = root.getChildCount();

        assertEquals(0, f1.getChildCount());
        assertEquals(0, f2.getChildCount());

        FolderUpdate up = new FolderUpdate(f2);
        up.setParentId(f1.getId());

        folderDao.update(f2.getId(), up);

        f1 = folderDao.get(f1.getId());
        assertEquals(1, f1.getChildCount());

        root = folderDao.get(getRootFolderId());
        assertEquals(rootCount -1, root.getChildCount());
    }

    @Test
    public void testDelete() {
        int count = folderDao.getChildren(getRootFolderId()).size();
        Folder f = folderDao.create(new FolderSpec("test"));
        assertTrue(folderDao.delete(f));
        assertEquals(count, folderDao.getChildren(getRootFolderId()).size());
    }

    @Test
    public void testCount() {
        int count = folderDao.count();
        Folder f = folderDao.create(new FolderSpec("test"));
        assertEquals(count+1, folderDao.count());
    }

    @Test
    public void testExists() {
        FolderSpec builder = new FolderSpec("foo");
        Folder folder1 = folderDao.create(builder);
        assertTrue(folderDao.exists(getRootFolderId(), "foo"));

        builder = new FolderSpec("bar", folder1);
        Folder folder2 = folderDao.create(builder);
        assertTrue(folderDao.exists(folder1.getId(), "bar"));
    }

    @Test
    public void testGetAllIds() {
        FolderSpec spec1 = new FolderSpec("foo");
        Folder folder1 = folderDao.create(spec1);

        DyHierarchySpec spec = new DyHierarchySpec().setFolderId(folder1.getId());
        spec.setLevels(
                ImmutableList.of(
                        new DyHierarchyLevel("source.type.raw"),
                        new DyHierarchyLevel("source.extension.raw")));
        DyHierarchy dyhi = dyHierarchyDao.create(spec);

        FolderSpec spec2 = new FolderSpec("bar");
        spec2.setDyhiId(dyhi.getId());
        Folder folder2 = folderDao.create(spec2);
        List<UUID> ids = folderDao.getAllIds(dyhi);
        assertTrue(ids.contains(folder2.getId()));
    }

    @Test
    public void testDeleteAllById() {

        FolderSpec spec1 = new FolderSpec("foo");
        Folder folder1 = folderDao.create(spec1);

        FolderSpec spec2 = new FolderSpec("bar");
        Folder folder2 = folderDao.create(spec2);

        assertEquals(2, folderDao.deleteAll(Lists.newArrayList(folder1.getId(), folder2.getId())));
    }
}

File: archivist/src/test/kotlin/com/zorroa/archivist/repository/FolderDaoTests.kt
package com.zorroa.archivist.repository


import com.google.common.collect.ImmutableList
import com.google.common.collect.Lists
import com.zorroa.archivist.AbstractTest
import com.zorroa.archivist.domain.*
import com.zorroa.archivist.security.hasPermission
import com.zorroa.common.domain.Access
import com.zorroa.security.Groups
import org.junit.Assert.*
import org.junit.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.dao.DataIntegrityViolationException
import java.io.IOException

class FolderDaoTests : AbstractTest() {

    @Autowired
    lateinit var taxonomyDao: TaxonomyDao

    @Autowired
    lateinit var folderDao: FolderDao

    @Autowired
    lateinit var dyHierarchyDao: DyHierarchyDao

    @Autowired
    lateinit var organizationDao: OrganizationDao

    @Test
    fun testCreateAndGet() {
        val name = "Foobar the folder"
        val builder = FolderSpec(name)
        val (id) = folderDao.create(builder)

        val (_, name1) = folderDao.get(id)
        assertEquals(name1, name)
    }

    @Test
    @Throws(IOException::class)
    fun testChildCount() {
        val name = "Foobar the folder"
        val builder1 = FolderSpec(name)
        var folder1 = folderDao.create(builder1)
        assertEquals(0, folder1.childCount.toLong())

        val builder2 = FolderSpec(name + "2", folder1)
        val folder2 = folderDao.create(builder2)
        folder1 = folderDao.get(folder1.id)
        assertEquals(1, folder1.childCount.toLong())

        folderDao.delete(folder2)
        folder1 = folderDao.get(folder1.id)
        assertEquals(0, folder1.childCount.toLong())
    }

    @Test
    @Throws(IOException::class)
    fun testGetByParentAndName() {
        val name = "test"
        val builder = FolderSpec(name)
        val folder1 = folderDao.create(builder)
        val folder2 = folderDao.get(folderDao.getRootFolder().id, name, false)
        assertEquals(folder1, folder2)
    }

    @Test
    @Throws(IOException::class)
    fun testGetRootFolder() {
        val folder = folderDao.getRootFolder()
        assertNull(folder.parentId)
        assertEquals("/", folder.name)
    }

    @Test
    @Throws(IOException::class)
    fun createRootFolder() {
        val org = organizationDao.create(OrganizationSpec("test"))
        val root = folderDao.createRootFolder(org)
        assertEquals("/", root.name)
        assertNull(root.parentId)
    }

    @Test
    @Throws(IOException::class)
    fun testGetAll() {
        val folder1 = folderDao.create(FolderSpec("test1"))
        val folder2 = folderDao.create(FolderSpec("test2"))
        val folder3 = folderDao.create(FolderSpec("test3"))
        val some = folderDao.getAll(listOf(folder2.id, folder3.id))
        assertTrue(some.contains(folder2))
        assertTrue(some.contains(folder3))
        assertFalse(some.contains(folder1))
        assertTrue(folderDao.getAll(listOf()).isEmpty())
    }

    @Test(expected = DataIntegrityViolationException::class)
    @Throws(IOException::class)
    fun testCreateDuplicate() {
        val name = "Foobar the folder"
        val builder = FolderSpec(name)
        folderDao.create(builder)
        folderDao.create(builder)
    }

    @Test
    fun testGetChildren() {
        val name = "Grandpa"
        var builder = FolderSpec(name)
        val grandpa = folderDao.create(builder)
        builder = FolderSpec("Dad", grandpa)
        val dad = folderDao.create(builder)
        builder = FolderSpec("Uncle", grandpa)
        val (id, name1, parentId, organizationId, dyhiId, user, timeCreated, timeModified, recursive, dyhiRoot, dyhiField, childCount, acl, search, taxonomyRoot, attrs) = folderDao.create(builder)
        builder = FolderSpec("Child", dad)
        val (id1, name2, parentId1, organizationId1, dyhiId1, user1, timeCreated1, timeModified1, recursive1, dyhiRoot1, dyhiField1, childCount1, acl1, search1, taxonomyRoot1, attrs1) = folderDao.create(builder)

        val folders = folderDao.getChildren(grandpa)
        assertEquals(2, folders.size.toLong())
    }

    @Test
    fun testsetDyHierarchyRoot() {
        val name = "Foobar the folder"
        val builder = FolderSpec(name)
        val folder = folderDao.create(builder)
        assertTrue(folderDao.setDyHierarchyRoot(folder, "source.extension"))
    }

    @Test
    fun testSetTaxonomyRoot() {
        val name = "Foobar the folder"
        val builder = FolderSpec(name)
        val folder = folderDao.create(builder)

        val tax = taxonomyDao!!.create(TaxonomySpec(folder))
        assertTrue(folderDao.setTaxonomyRoot(folder, true))
        assertFalse(folderDao.setTaxonomyRoot(folder, true))

        assertTrue(folderDao.setTaxonomyRoot(folder, false))
        assertFalse(folderDao.setTaxonomyRoot(folder, false))
    }

    @Test
    fun removeDyHierarchyRoot() {
        val name = "Foobar the folder"
        val builder = FolderSpec(name)
        val folder = folderDao.create(builder)
        assertTrue(folderDao.setDyHierarchyRoot(folder, "source.extension"))
        assertTrue(folderDao.removeDyHierarchyRoot(folder))
    }


    @Test
    fun testGetChildrenInsecure() {
        authenticate("user")
        assertFalse(hasPermission(Groups.ADMIN))

        val pub = folderDao.get(folderDao.getRootFolder().id, "Users", false)
        val startCount = folderDao.getChildren(pub.id).size
        val (id, name, parentId, organizationId, dyhiId, user, timeCreated, timeModified, recursive, dyhiRoot, dyhiField, childCount, acl, search, taxonomyRoot, attrs) = folderDao.create(FolderSpec("level1", pub))
        val (id1, name1, parentId1, organizationId1, dyhiId1, user1, timeCreated1, timeModified1, recursive1, dyhiRoot1, dyhiField1, childCount1, acl1, search1, taxonomyRoot1, attrs1) = folderDao.create(FolderSpec("level2", pub))
        val f3 = folderDao.create(FolderSpec("level3", pub))
        folderDao.setAcl(f3.id, Acl().addEntry(
                permissionService.getPermission(Groups.ADMIN)))

        assertFalse(folderDao.hasAccess(f3, Access.Read))
        assertEquals(6, folderDao.getChildrenInsecure(pub.id).size.toLong())
        assertEquals((startCount + 2).toLong(), folderDao.getChildren(pub.id).size.toLong())
    }

    @Test
    @Throws(IOException::class)
    fun testHasAccess() {
        authenticate("user")
        val builder = FolderSpec("test")
        val folder1 = folderDao.create(builder)
        assertTrue(folderDao.hasAccess(folder1, Access.Read))

        val p = permissionService.createPermission(PermissionSpec("group", "foo"))
        val acl = Acl()
        acl.addEntry(p, Access.Read)
        folderDao.setAcl(folder1.id, acl)

        assertFalse(folderDao.hasAccess(folder1, Access.Read))
    }

    @Test
    @Throws(IOException::class)
    fun testGetAndSetAcl() {
        val builder = FolderSpec("test")
        val (id) = folderDao.create(builder)

        val p = permissionService.createPermission(PermissionSpec("group", "foo"))
        folderDao.setAcl(id, Acl().addEntry(p, Access.Read))

        val acl = folderDao.getAcl(id)
        assertTrue(acl.hasAccess(p.id, Access.Read))
    }

    @Test
    @Throws(IOException::class)
    fun testGetAndUpdateAcl() {
        val builder = FolderSpec("test")
        val (id) = folderDao.create(builder)

        val p1 = permissionService.createPermission(PermissionSpec("group", "foo"))
        folderDao.setAcl(id, Acl().addEntry(p1, Access.Read))

        var acl = folderDao.getAcl(id)
        assertTrue(acl.hasAccess(p1.id, Access.Read))

        val p2 = permissionService.createPermission(PermissionSpec("group", "bar"))
        folderDao.updateAcl(id, Acl().addEntry(p2, Access.Read))

        acl = folderDao.getAcl(id)
        assertTrue(acl.hasAccess(p1.id, Access.Read))
        assertTrue(acl.hasAccess(p2.id, Access.Read))
    }

    @Test
    @Throws(IOException::class)
    fun testUpdateExitingAcl() {
        val builder = FolderSpec("test")
        val (id) = folderDao.create(builder)

        val p1 = permissionService.createPermission(PermissionSpec("group", "foo"))
        folderDao.setAcl(id, Acl().addEntry(p1, Access.Read))
        folderDao.updateAcl(id, Acl().addEntry(p1, Access.Read, Access.Write))

        val acl = folderDao.getAcl(id)
        assertTrue(acl.hasAccess(p1.id, Access.Read))
        assertTrue(acl.hasAccess(p1.id, Access.Write))
        assertFalse(acl.hasAccess(p1.id, Access.Export))
    }

    @Test(expected = IllegalArgumentException::class)
    @Throws(IOException::class)
    fun testGetAndSetBadAcl() {
        val builder = FolderSpec("test")
        val (id) = folderDao.create(builder)

        val p = permissionService.createPermission(PermissionSpec("group", "foo"))
        val acl = Acl()
        acl.add(AclEntry(p.id, 12))
        folderDao.setAcl(id, acl)
    }

    @Test
    fun testUpdate() {
        val spec = FolderSpec("v1")
        val v1 = folderDao.create(spec)

        val up = FolderUpdate(v1)
        up.name = "v2"

        folderDao.update(v1.id, up)
        val (_, name, _, _, _, _, _, _, _, _, _, _, acl, search) = folderDao.get(v1.id)
        assertEquals("v2", name)
        assertEquals(null, search)
        assertEquals(0, acl!!.size.toLong())
    }

    @Test
    fun testUpdateWithMove() {
        val spec1 = FolderSpec("f1")
        var f1 = folderDao.create(spec1)

        val spec2 = FolderSpec("f2")
        val f2 = folderDao.create(spec2)

        var root = folderDao.get(folderDao.getRootFolder().id)
        val rootCount = root.childCount

        assertEquals(0, f1.childCount.toLong())
        assertEquals(0, f2.childCount.toLong())

        val up = FolderUpdate(f2)
        up.parentId = f1.id

        folderDao.update(f2.id, up)

        f1 = folderDao.get(f1.id)
        assertEquals(1, f1.childCount.toLong())

        root = folderDao.get(folderDao.getRootFolder().id)
        assertEquals((rootCount - 1).toLong(), root.childCount.toLong())
    }

    @Test
    fun testDelete() {
        val count = folderDao.getChildren(folderDao.getRootFolder().id).size
        val f = folderDao.create(FolderSpec("test"))
        assertTrue(folderDao.delete(f))
        assertEquals(count.toLong(), folderDao.getChildren(folderDao.getRootFolder().id).size.toLong())
    }

    @Test
    fun testCount() {
        val count = folderDao.count()
        val (id, name, parentId, organizationId, dyhiId, user, timeCreated, timeModified, recursive, dyhiRoot, dyhiField, childCount, acl, search, taxonomyRoot, attrs) = folderDao.create(FolderSpec("test"))
        assertEquals((count + 1).toLong(), folderDao.count().toLong())
    }

    @Test
    fun testExists() {
        var builder = FolderSpec("foo")
        val folder1 = folderDao.create(builder)
        assertTrue(folderDao.exists(folderDao.getRootFolder().id, "foo"))

        builder = FolderSpec("bar", folder1)
        val (id, name, parentId, organizationId, dyhiId, user, timeCreated, timeModified, recursive, dyhiRoot, dyhiField, childCount, acl, search, taxonomyRoot, attrs) = folderDao.create(builder)
        assertTrue(folderDao.exists(folder1.id, "bar"))
    }

    @Test
    fun testGetAllIds() {
        val spec1 = FolderSpec("foo")
        val (id) = folderDao.create(spec1)

        val spec = DyHierarchySpec().setFolderId(id)
        spec.levels = ImmutableList.of(
                DyHierarchyLevel("source.type.raw"),
                DyHierarchyLevel("source.extension.raw"))
        val dyhi = dyHierarchyDao.create(spec)

        val spec2 = FolderSpec("bar")
        spec2.dyhiId = dyhi.id
        val (id1) = folderDao.create(spec2)
        val ids = folderDao.getAllIds(dyhi)
        assertTrue(ids.contains(id1))
    }

    @Test
    fun testDeleteAllById() {

        val spec1 = FolderSpec("foo")
        val (id) = folderDao.create(spec1)

        val spec2 = FolderSpec("bar")
        val (id1) = folderDao.create(spec2)

        assertEquals(2, folderDao.deleteAll(Lists.newArrayList(id, id1)).toLong())
    }
}
